### PR TITLE
Make Client configurable

### LINKED ISSUE #539: Make Client configurable with Layers
`kube::Client` now takes `Service<http::Request<hyper::Body>>`, so users can create a client with custom Service. However, to make this useful beyond simple testing, we need to provide methods to create parts of the default Service (mainly TLS and auth providers).

The goal is to allow full customization at both `connector` and `service` levels:
```rust
let https = HttpsConnector::from((http, tls));
let connector = TimeoutConnector::new(https, timeout);
let https_client = hyper::Client::builder()
    .build(connector);

let service = tower::ServiceBuilder::new()
    .layer(set_cluster_url)
    .layer(add_authorization)
    .layer(decompression)
    .layer(trace_layer)
    .service(https_client);
let client = kube::Client::new(service);
```
We should also use layers from [`tower-http`](https://github.com/tower-rs/tower-http) whenever possible.

### TLS

Not having this makes it impossible for custom client to communicate with real clusters.
We currently do the steps described below, and we need to provide a way to create the necessary struct from
`kube::Config` (step 1).

#### Native TLS

1. Create `native_tls::TlsConnector` from `config.identity` and `config.root_cert` (+ `config.accept_invalid_certs`)
2. `hyper_tls::HttpsConnector::from((http, tls))`
    - `http` is `hyper::client::HttpConnector` and `tls` is `tokio_native_tls::TlsConnector`
    - `tokio_native_tls::TlsConnector` is created from `native_tls::TlsConnector`

> Note that 1 depends on `openssl` to create PKCS12 archive for `native_tls::Identity`.
> We should replace that part with pure Rust so `openssl` isn't necessary on macOS and Windows.
> It's currently painful to do CI on Windows.

#### Rustls

1. Create `rustls::ClientConfig` from `config.identity` and `config.root_cert` (+ `config.accept_invalid_certs`)
2. `hyper_rustls::HttpsConnector::from((http, config))`
    - `http` is `hyper::client::HttpConnector` and `config` is `Arc<rustls::ClientConfig>`

### Authorization

Authorization with Basic/Bearer are easy to implement, but refreshable token is pretty complex. We should make that public.

#### Basic

Users can use `AddAuthorizationLayer::basic(user, pass)` from `tower_http`.
Or use `map_request` to add `Authorization` header.

#### Bearer

Users can use `AddAuthorizationLayer::bearer(token)` from `tower_http`.
Or use `map_request` to add `Authorization` header.

#### Refreshable (exec and oauth)

`kube::service::AuthLayer` should be renamed and made public.
This layer refreshes the token when necessary before setting the header.

### Cluster URL

It's not difficult to implement, but we should provide a convenient way to set cluster URL because it's almost always necessary. ~~`SetClusterUrlLayer::new(cluster_url)`.~~

`SetBaseUrlLayer`. Probably general enough to ask to add to `tower_http`.

### Decompression Support

Use `DecompressionLayer` from `tower_http`.

### Log/Trace

Use `TraceLayer` from `tower_http`.

### Headers

We should remove `headers` from `Config`. Users can use a layer to add them.

## TODO

- [x] Add methods `Config::native_tls_connector` and `Config::rustls_client_config` for users to configure TLS connection when building a custom client (#540)
- [x] Remove the restriction of requiring one of `native-tls` or `rustls-tls` when `client` feature is enabled. Allow one, both or none. When both, the default Service will use `native-tls` because of #153. (#540)
- [x] Remove TLS features from `kube-runtime` (#540)
- [x] Add public `SetBaseUriLayer` (#540)
- [x] Add `ConfigExt::auth_layer` to return optional layer that manages `Authorization` header (#540)
- [x] Use `AddAuthorization::basic`
- [x] Use `AddAuthorization::bearer`
- [x] Use [`DecompressionLayer`](https://docs.rs/tower-http/0.1.0/tower_http/decompression/index.html) instead of custom decompression module
- [x] Use [`TraceLayer`](https://docs.rs/tower-http/0.1.0/tower_http/trace/index.html) instead of `LogRequest`
- [x] Remove `headers` from `Config`
- Open PR to add `kube` to `tower`'s README https://github.com/tower-rs/tower/tree/master/tower#library-support

---

@clux Just summarizing what I have in mind and what needs to be done. Let me know what you think. Hoping to make some time to work on this soon, but if not, I can at least review PRs.

### PR DESCRIPTION
**Pull request by kazk**
Customize client with [Layers](https://docs.rs/tower/0.4.8/tower/layer/trait.Layer.html):
```rust
use kube::{client::ConfigExt, Client, Config};
use tower::ServiceBuilder;
use tower_http::{decompression::DecompressionLayer, trace::TraceLayer};

let config = Config::infer().await?;
let https = config.native_tls_https_connector()?;
let client = Client::new(
    ServiceBuilder::new()
        .layer(config.base_uri_layer())
        .option_layer(config.auth_layer()?)
        .layer(DecompressionLayer::new())
        .layer(TraceLayer::new_for_http())
        .service(hyper::Client::builder().build(https)),
);
```

Customize `HttpsConnector`:
```rust
let https = ServiceBuilder::new()
    .layer(layer_fn(|c| {
        let mut conn = TimeoutConnector::new(c);
        conn.set_connect_timeout(Some(Duration::from_secs(5)));
        conn
    }))
    .service(config.native_tls_https_connector()?);
```

It's also possible to create `HttpsConnector` from `TlsConnector`:
```rust
let https = {
    let tls = tokio_native_tls::TlsConnector::from(
        config.native_tls_connector()?
    );
    let mut http = HttpConnector::new();
    http.enforce_http(false);
    HttpsConnector::from((http, tls))
};
```

## TLS Enhancements

- Add `kube::client::ConfigExt` extending `Config` for custom `Client`. This includes methods to configure TLS connection when building a custom client (#539)
  - `native-tls`: `Config::native_tls_https_connector` and `Config::native_tls_connector`
  - `rustls-tls`: `Config::rustls_https_connector` and `Config::rustls_client_config`
- Remove the requirement of having `native-tls` or `rustls-tls` enabled when `client` is enabled. Allow one, both or none.
  - When both, the default Service will use `native-tls` because of #153. `rustls` can be still used with a custom client. Users will have an option to configure TLS at runtime.
  - When none, HTTP connector is used.
- Remove TLS features from `kube-runtime`
  - **BREAKING**: Features must be removed if specified 
- Remove `client` feature from `native-tls` and `rust-tls` features
  - `config` + `native-tls`/`rustls-tls` can be used independently, e.g., to create a simple HTTP client
  - **BREAKING**: `client` feature must be added if `default-features = false`

## Layers

- `ConfigExt::base_uri_layer` (`BaseUriLayer`) to set cluster URL (#539)
- `ConfigExt::auth_layer` that returns optional layer to manage `Authorization` header (#539)
- `gzip`: Replaced custom decompression module with [`DecompressionLayer`](https://docs.rs/tower-http/0.1.0/tower_http/decompression/index.html) from `tower-http` (#539)
- Replaced custom `LogRequest` with [`TraceLayer`](https://docs.rs/tower-http/0.1.0/tower_http/trace/index.html) from `tower-http` (#539)
  - Request body is no longer shown
- Basic and Bearer authentication using `AddAuthorizationLayer` (borrowing from https://github.com/tower-rs/tower-http/pull/95 until released)

## Changes

- **BREAKING**: Remove `headers` from `Config`. It was originally added to allow injecting arbitrary headers, but now a custom client can be used to do that.
- Add support for loading proxy URL from kube config. Proxy itself is not supported yet.

## Dependency Changes

- Remove `static_assertions` since it's no longer used
- Replace `tokio_rustls` with `rustls` and `webpki` since we're not using `tokio_rustls` directly
- Replace uses of `rustls::internal::pemfile` with `rustls-pemfile`
- Remove `url` and always use `http::Uri`
  - **BREAKING**: `Config::cluster_url` is now `http::Uri`
  - **BREAKING**: `Error::InternalUrlError(url::ParseError)` and `Error::MalformedUrl(url::ParseError)` replaced by `Error::InvalidUri(http::uri::InvalidUri)`

---

Closes #539

### CODE REVIEW

File: kube/src/lib.rs
Comment (kazk) [inline kube/src/lib.rs:115]: Unrelated to this PR, but I found one downside of `cfg_client!`.

`cargo +nightly fmt --all` doesn't work anymore :( Apparently it doesn't look inside macros. `rustfmt +nightly --edition 2018 **/*/*.rs` works.

File: kube/src/client/mod.rs
Comment (kazk) [inline kube/src/client/mod.rs:91]: Took me a few days to come up with this, but should be worth it. This change allows us to take advantage of the Tower ecosystem much more without introducing type parameters.

Shouldn't be a breaking change, but need to test more to make sure.

File: examples/pod_api.rs
Comment (kazk) [inline examples/pod_api.rs:13]: Run this example to see the default outputs from `TraceLayer` which replaced `LogRequest`.
It's fully customizable as [documented](https://docs.rs/tower-http/0.1.0/tower_http/trace/index.html), but kept it simple.

File: examples/pod_api.rs
Comment (kazk) [inline examples/pod_api.rs:13]: This example can be used to see the output from `TraceLayer` for the default client.

image

File: examples/custom_client.rs
Comment (kazk) [inline examples/custom_client.rs:32]: A minimal `Client` can be created like this from `Config`.

File: examples/pod_api.rs
Comment (clux) [reply examples/pod_api.rs:13]: I think that looks great, and probably sufficient to rip out all the `#[instrumen]` instructions in the `Api`.

One philosophical point; having the duration as a separate log line does feel to me like it's encroaching on what should be tracing's turf:
Screenshot from 2021-06-02 13-45-20
A span viewer will always have the timing information anyway, whereas a line for it here is just a harder way to read this, and it imo pollutes the log space.

File: examples/Cargo.toml
Comment (clux) [inline examples/Cargo.toml:53]: :tada: :tada: :tada:

File: examples/custom_client_trace.rs
Comment (clux) [inline examples/custom_client_trace.rs:59]: This looks very nice.

File: examples/custom_client_tls.rs
Comment (clux) [inline examples/custom_client_tls.rs:42]: This seems like a strange option in `HttpConnector` to have default-true, when you can turn it into an `HttpsConnector` :thinking:

File: examples/custom_client_trace.rs
Comment (clux) [inline examples/custom_client_trace.rs:41]: maybe link the semantic convention doc here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#http-client

File: examples/custom_client_trace.rs
Comment (clux) [inline examples/custom_client_trace.rs:112]: Just for speed of testing + readability, maybe we should have a simpler, less wait-based rest in these 3 custom_client* examples. Something like a two line: `Api<ConfigMap>` list in a namespace. The interesting part of the example here is the top bit :-)

File: kube-runtime/Cargo.toml
Comment (clux) [inline kube-runtime/Cargo.toml:35]: RIP infectious features :skull:

File: kube/src/client/body.rs
Comment (clux) [inline kube/src/client/body.rs:25]: This seems reasonable, pretty light weight. It looks like there were even thoughts of doing this in `http-body` at some point: https://github.com/hyperium/http-body/pull/36

File: kube/src/client/mod.rs
Comment (clux) [inline kube/src/client/mod.rs:447]: While I am happy that we can do without `static_assertions`, I'm a little sad that the build-time; one-tls-stack-dependency guarantee goes away. But it does seem to be the popular demand (even if you end up with bigger executables), and managing build deps can be a complex problem by itself.

However, this does mean that if rustls gets their dns game together (that one caveat listed for rustls in the readme), and we consider seriously switching to rustls as the default, then we now have a new breaking change for it.

I think it's still probably sensible though. It does mean we can do all-features builds again, and generate docs for that.

File: kube/src/client/mod.rs
Comment (clux) [inline kube/src/client/mod.rs:485]: You already got the required ones there! Although `http.uri` should be `http.url` according to semantic_conventions + otel.

The `otel.kind` + `otel.status_code` should probably be set as well (they are displayed by default in at least some trace viewers, see screenshot linked in other comment).

File: kube/src/client/mod.rs
Comment (clux) [inline kube/src/client/mod.rs:491]: I would take out the `debug!` line here, or downgrade them to `trace!` level.

**Review by clux — COMMENTED** [review]
looks great. just small philosophical comments. have not dug through the tls stuff much yet.

File: kube/src/client/mod.rs
Comment (clux) [reply kube/src/client/mod.rs:485]: shame there's no way to get the `HTTP` name to be more dynamic. If we got the name of the request call (like `patch_status`) to be the span name, then we could possibly drop the manual `#[instrument]`ations.

File: examples/custom_client_tls.rs
Comment (kazk) [reply examples/custom_client_tls.rs:42]: Yeah, it looks awkward. But `HttpConnector` is independent and enforcing `http` protocol by default prevents misuse with other protocols like `file://`. I just wish there was a constructor or a builder.

File: kube/src/client/mod.rs
Comment (kazk) [reply kube/src/client/mod.rs]: Yeah, it's possible to derive a name from `Request` and use that with `otel.name` to override the span name:

```rust
TraceLayer::new_for_http()
    .make_span_with(|req: &Request<hyper::Body>| {
        tracing::debug_span!(
            "HTTP",
            http.method = %req.method(),
            http.url = %req.uri(),
            http.status_code = tracing::field::Empty,
            // Override the span name sent to OpenTelemetry exporters.
            otel.name = otel_name(req),
            otel.kind = "client",
            otel.status_code = tracing::field::Empty,
        )
    })
```

But not sure what's good.

We can do `HTTP {METHOD_NAME}` like suggested in the semantic convention.
```rust
#[inline]
fn otel_name(req: &Request<hyper::Body>) -> &'static str {
    match *req.method() {
        http::Method::POST => "HTTP POST",
        http::Method::GET => "HTTP GET",
        http::Method::PUT => "HTTP PUT",
        http::Method::PATCH => "HTTP PATCH",
        http::Method::DELETE => "HTTP DELETE",
        _ => "HTTP",
    }
}
```

> If we got the name of the request call (like `patch_status`)

Maybe insert the name in [extensions](https://docs.rs/http/0.2.4/http/struct.Extensions.html):
```rust
// in Api
req.extensions_mut().insert("patch_status");
```
and
```rust
otel.name = req.extensions().get::<&'static str>().unwrap_or(&"HTTP"),
```
?

File: kube/src/client/mod.rs
Comment (kazk) [reply kube/src/client/mod.rs:429]: Removing TLS stack requirement is necessary to let users build libraries on top of `kube`. Otherwise, they'll end up having the awkward TLS features like `kube-runtime` did.

I honestly don't know how users using `kube` directly can end up with both enabled _accidentally_ which `static_assertions` prevents. Also, having `native-tls` by default and requiring `default-features = false` is very common. If users are accidentally enabling both often, then we should improve the documentation.

> we consider seriously switching to rustls as the default

I do want rustls to become the default eventually, but I don't think this will happen any time soon. Validating certificates for IP addresses is not the only issue. Support for different types of private keys are lacking. For example, EC Private Key used by k3d is not supported, so rustls fails to find the key (only PKCS#8 and RSA (PKCS#1) can be extracted from PEM). To use rustls with k3d, we need to change the cluster's host to `localhost` and convert the private key in kubeconfig to PKCS#8 using OpenSSL. We can do this in `kube`, but OpenSSL will be required :(

Native TLS connector only supports loading Identity from PKCS#12, but we're using OpenSSL to create the archive, so the conversions is done there. This is also not ideal because of OpenSSL dependency on macOS and Windows, especially the latter is painful. It's possible to create PKCS#12 archive in pure Rust, but some key formats won't be supported.

File: kube/src/client/mod.rs
Comment (clux) [reply kube/src/client/mod.rs]: Oh nice! Extensions is a good idea! Interesting solution they have chosen with the TypeId map rather than explicit keys, but seems to be a memory optimisation for small amount of entries. It's probably okay to use it, it generally stays internal anyway (unless people pick apart it with our `Request` type, but that's also fine).

File: kube/src/client/mod.rs
Comment (clux) [reply kube/src/client/mod.rs:430]: Yeah, I definitely think this is the right thing to do.

For some of these rustls caveats, it might be worth creating another tracking issue so that this is not lost. I thought it was only the IP issue :-(

Comment (kazk) [reply kube/src/client/mod.rs:430]: Opened #542

File: kube/src/client/config_ext.rs
Comment (clux) [inline kube/src/client/config_ext.rs:13]: This seems to be a variably-sized trait based on features. That looks difficult to implement outside of kube. Is the plan for this to become public at some point?

Comment (kazk) [reply kube/src/client/config_ext.rs:13]: No, I'll seal it and document it.

File: kube/src/client/mod.rs
Comment (clux) [reply kube/src/client/mod.rs]: New version looks great. I just tried running through Tempo via controller-rs.

Screenshot from 2021-06-04 22-21-41

Semantic convention looks accurate, extension passed span name is picked up correctly.

Interestingly, if you are in tracing land, then the `debug!("requesting")` is pointless and only inflates the trace, but if you are in stdout land, then you need to have it - as `tracing_subscriber::fmt::init()` only prints events and not spans. Don't think there's anything that can be done about that at the moment though.

File: examples/custom_client.rs
Comment (clux) [inline examples/custom_client.rs:20]: nit: the name `set_server_layer` reads to me like it's a layer for server operation, which, while a bit nonsensical here, still makes me have to go to the docs. Maybe `base_url_layer()` (with the `set_` verb implicit) is more intuitive here?

File: kube/src/client/auth/refreshing_token.rs
Comment (clux) [inline kube/src/client/auth/refreshing_token.rs:14]: `RefreshTokenLayer` is probably ok.
`RefreshingTokenLayer` is funny though; like, wow, very original take on a token layer.

File: kube/src/client/config_ext.rs
Comment (clux) [inline kube/src/client/config_ext.rs:17]: I didn't know about this pattern, but am a fan. Feels like a very smart way to avoid having all the public methods in a big namespace :+1:

File: kube/src/client/middleware/base_uri.rs
Comment (clux) [inline kube/src/client/middleware/base_uri.rs:9]: Think this one should just be named `BaseUriLayer` for consistency. Our core exported layers are all without verbs (except for the temporary `AddAuthorizationLayer` taken from the linked tower-http pr), and having that consistency - where possible - feels cleaner to me.

NB: We might not be able to go for a verb-less style in general / long-term, as [tower-http](https://docs.rs/tower-http/0.1.0/tower_http/) does uncover legit cases where they need more than one verb on the same resource (`SetResponseHeaderLayer` vs `PropagateHeaderLayer`), but they also drop the verb when there's no disambiguation needed.

**Review by clux — APPROVED** [review]
Looks amazing. Huge work :100: 

Had another go over. TLS looks all great, good error handling everywhere, docs everywhere, things with TODOs are all pub(crate).

Left (another) philosiphical naming nit, but beyond this I am all good with this.

**Comment by clux** [comment]
I'll leave it to you to make the final call on naming, and will make a release after you merge this :-)

File: kube/src/client/config_ext.rs
Comment (kazk) [reply kube/src/client/config_ext.rs:17]: Yeah, the downside is the extra import, but I think the explicit `kube::client::ConfigExt` actually makes sense for this. I also wanted to keep client stuff out of the config module.

**Comment by clux** [comment]
Release in 0.56.0! Thank you so much!

File: kube/src/client/mod.rs
Comment (kazk) [reply kube/src/client/mod.rs]: > as `tracing_subscriber::fmt::init()` only prints events and not spans. 

It's possible to make `tracing_subscriber::fmt` log events on span lifecycle.

```rust
use tracing_subscriber::fmt::format::FmtSpan;

tracing_subscriber::fmt()
    .with_span_events(FmtSpan::CLOSE)
    .init();
```
will log when the span closes.

**Pull request merged**
